### PR TITLE
[TISPARK-102] Use reflect to support multi-spark-version rather than profile

### PR DESCRIPTION
**Pull request by birdstorm**

### CODE REVIEW

**Comment by birdstorm** [comment]
we may first check if it works locally.

File: core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala [hunk 1]
```diff
@@ -57,53 +57,57 @@ case class CoprocessorRDD(output: Seq[Attribute], tiRdd: TiRDD) extends LeafExec
...
-    //      isOrderSensitive: Boolean = false): RDD[U]
-    //
-    // Hereby we use reflection to support different Spark versions.
-    val method = try {
+  // In Spark 2.3.0 the method declaration is:
+  // private[spark] def mapPartitionsWithIndexInternal[U: ClassTag](
+  //      f: (Int, Iterator[T]) => Iterator[U],
+  //      preservesPartitioning: Boolean = false): RDD[U]
+  //
+  // In other Spark 2.3.x versions, the method declaration is:
```
Comment (marsishandsome) [inline core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala]: x > 0

File: core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala [hunk 2]
```diff
@@ -57,53 +57,57 @@ case class CoprocessorRDD(output: Seq[Attribute], tiRdd: TiRDD) extends LeafExec
...
+    classOf[RDD[InternalRow]]
+      .getDeclaredMethod(
+        "mapPartitionsWithIndexInternal",
+        classOf[(Int, Iterator[InternalRow]) => Iterator[UnsafeRow]],
+        classOf[Boolean],
+        classOf[Boolean],
+        classOf[ClassTag[UnsafeRow]]
+      )
+  } catch {
+    case _: Throwable =>
```
Comment (marsishandsome) [inline core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala]: 1. it's better to provide a general function to do spark version related reflection
2. you can read spark version from `org.apache.spark.SPARK_VERSION`, and then call the correct reflected function. Do not need to use try catch.

Comment (birdstorm) [reply core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala]: agree.

File: core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala (hunk 1 above)
Comment (birdstorm) [reply core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala]: word `other` excluded 2.3.0

**Pull request merged**